client/web/src/repo/blob/codemirror/tooltips/CodeIntelTooltip.tsx

```diff
@@ -0,0 +1,95 @@
+import { EditorView, Tooltip, TooltipView } from '@codemirror/view'
+import { of } from 'rxjs'
+
+import { HoverMerged } from '@sourcegraph/client-api'
+import { MarkupKind } from '@sourcegraph/extension-api-classes'
+import { ActionItemAction } from '@sourcegraph/shared/src/actions/ActionItem'
+import { Occurrence } from '@sourcegraph/shared/src/codeintel/scip'
+import { BlobViewState, toPrettyBlobURL } from '@sourcegraph/shared/src/util/url'
+
+import { HovercardView, HoverData } from '../hovercard'
+import { rangeToCmSelection } from '../occurrence-utils'
+import { blobInfoFacet, textDocumentImplemenationSupport } from '../token-selection/facets'
+import { modifierClickDescription } from '../token-selection/modifier-click'
+
+export interface HoverResult {
+    markdownContents: string
+    hoverMerged?: HoverMerged | null
+    isPrecise?: boolean
+}
+
+// CodeMirror tooltip wrapper around the "code intel" popover.  Implemented as a
+// class so that we can detect it with instanceof checks. This class
+// reimplements logic from `getHoverActions` in
+// 'client/shared/src/hover/actions.ts' because that function is difficult to
+// reason about and has surprising behavior.
+export class CodeIntelTooltip implements Tooltip {
+    public readonly above = true
+    public readonly pos: number
+    public readonly end: number
+    public readonly create: () => TooltipView
+    constructor(view: EditorView, occurrence: Occurrence, hover: HoverResult) {
+        const { markdownContents } = hover
+        const range = rangeToCmSelection(view.state, occurrence.range)
+        this.pos = range.from
+        this.end = range.to
+        this.create = () => {
+            const blobInfo = view.state.facet(blobInfoFacet)
+            const referencesURL = toPrettyBlobURL({
+                ...blobInfo,
+                range: occurrence.range.withIncrementedValues(),
+                viewState: 'references',
+            })
+            const actions: ActionItemAction[] = [
+                {
+                    active: true,
+                    action: {
+                        id: 'findReferences',
+                        title: 'Find references',
+                        command: 'open',
+                        commandArguments: [referencesURL],
+                    },
+                },
+            ]
+            if (hover.isPrecise && view.state.facet(textDocumentImplemenationSupport)) {
```

Comment:
```suggestion
            if (hover.isPrecise && view.state.facet(blobPropsFacet).blobInfo.mode) {
```
I wonder if we can make use of existing `blobPropsFacet`, WDYT?